singularity-eos/eos/eos_mgusup.hpp

```diff
@@ -0,0 +1,365 @@
+                              Real *lambda = nullptr) const;
+  // Generic functions provided by the base class. These contain e.g. the vector
+  // overloads that use the scalar versions declared here
+  SG_ADD_BASE_CLASS_USINGS(MGUsup)
+  PORTABLE_INLINE_FUNCTION
+  int nlambda() const noexcept { return 0; }
+  static constexpr unsigned long PreferredInput() { return _preferred_input; }
+  static inline unsigned long scratch_size(std::string method, unsigned int nelements) {
+    return 0;
+  }
+  static inline unsigned long max_scratch_size(unsigned int nelements) { return 0; }
+  PORTABLE_INLINE_FUNCTION void PrintParams() const {
+    static constexpr char st[]{"MGUsup Params: "};
+    printf("%s rho0:%e T0:%e Cs:%e s:%e\n  G0:%e Cv0:%e E0:%e S0:%e\n", st, _rho0, _T0,
+           _Cs, _s, _G0, _Cv0, _E0, _S0);
+    printf("\n\n");
+  }
+  // Density/Energy from P/T not unique, if used will give error
+  PORTABLE_INLINE_FUNCTION void
+  DensityEnergyFromPressureTemperature(const Real press, const Real temp, Real *lambda,
+                                       Real &rho, Real &sie) const;
+  inline void Finalize() {}
+  static std::string EosType() { return std::string("MGUsup"); }
+  static std::string EosPyType() { return EosType(); }
+
+ private:
+  static constexpr const unsigned long _preferred_input =
+      thermalqs::density | thermalqs::specific_internal_energy;
+  Real _rho0, _T0, _Cs, _s, _G0, _Cv0, _E0, _S0;
+  void CheckMGUsup();
+};
+
+inline void MGUsup::CheckMGUsup() {
+
+  if (_rho0 < 0.0) {
+    PORTABLE_ALWAYS_THROW_OR_ABORT("Required MGUsup model parameter rho0 < 0");
+  }
+  if (_T0 < 0.0) {
+    PORTABLE_ALWAYS_THROW_OR_ABORT("Required MGUsup model parameter T0 < 0");
+  }
+  if (_Cs < 0.0) {
+    PORTABLE_ALWAYS_THROW_OR_ABORT("Required MGUsup model parameter Cs < 0");
+  }
+  if (_s < 0.0) {
+    PORTABLE_ALWAYS_THROW_OR_ABORT("Required MGUsup model parameter s < 0");
+  }
+  if (_G0 < 0.0) {
+    PORTABLE_ALWAYS_THROW_OR_ABORT("Required MGUsup model parameter G0 < 0");
+  }
+  if (_Cv0 < 0.0) {
+    PORTABLE_ALWAYS_THROW_OR_ABORT("Required MGUsup model parameter Cv0 < 0");
+  }
+}
+
+PORTABLE_INLINE_FUNCTION Real MGUsup::HugPressureFromDensity(Real rho) const {
+  Real eta = 1.0 - robust::ratio(_rho0, rho);
+  return _Cs * _Cs * _rho0 * robust::ratio(eta, (1.0 - _s * eta) * (1.0 - _s * eta));
+}
+PORTABLE_INLINE_FUNCTION Real MGUsup::HugInternalEnergyFromDensity(Real rho) const {
+  Real eta = 1.0 - robust::ratio(_rho0, rho);
+  return _E0 + robust::ratio(eta, 2.0 * _rho0) * HugPressureFromDensity(rho);
+}
+PORTABLE_INLINE_FUNCTION Real MGUsup::HugTemperatureFromDensity(Real rho) const {
+  int sumkmax = 20;
+  Real cutoff = 1.e-16;
+  Real eta = 1.0 - robust::ratio(_rho0, rho);
+  Real f1 = 1.0 - _s * eta;
+  if (f1 <= 0.0) {
+    PORTABLE_ALWAYS_THROW_OR_ABORT("MGUsup model parameters s and rho0 together with rho "
+                                   "give a negative argument for a logarithm.");
+  }
```

Comment:
Hmm. Related to the discussion with @jhp-lanl  above, I would prefer if there was a way to enforce that the code return something reasonable here, even with bad input... This can be a discussion for later, though, when people are actually using the model. 

I just worry that we are likely to hit this bound in real simulations and we don't want somebody's run to crash because of this. Maybe we just bound f1 > 0?

Reply:
We need to discuss this. It cannot be tested at initialization since \rho is varying, and thus \eta and thus s*\eta. A maximum \eta can be calculated of course (\eta_max = 1/s) and with that max Hugoniot pressure and max Hugoniot temperature. If this happens they should really not use these EOSs but use the PowerMG that I will implement next.

Reply:
@aematts I'm happy to discuss this here or Mattermost but if you'd like to meet, I'm happy to do that too. Just let me know

Reply:
Continued on Mattermost